drv/gimlet-hf-server/src/main.rs

```diff
@@ -180,6 +269,190 @@ impl ServerImpl {
+            // `dev_select_pin` is `Some(...)`, so we can unwrap returns from
+            // `self.set_dev(...)`.
+
+            // Look at the inactive slot first
+            self.set_dev(!prev_slot).unwrap();
+            let (a, _) = self.persistent_data_scan();
+
+            // Then switch back to our current slot, so that the resulting state
+            // is unchanged.
+            self.set_dev(prev_slot).unwrap();
+            let (b, _) = self.persistent_data_scan();
+
+            match (a, b) {
+                (Some(a), Some(b)) => {
+                    if a.epoch > b.epoch {
+                        Some(a)
+                    } else {
+                        Some(b)
+                    }
+                }
+                (Some(_), None) => a,
+                (None, Some(_)) => b,
+                (None, None) => None,
+            }
+        } else {
+            // The single-chip case is easy:
+            self.persistent_data_scan().0
+        };
+
+        best.ok_or(HfError::NoPersistentData)
+    }
+
+    /// Scans the currently active flash IC for persistent data
+    ///
+    /// Returns a tuple containing
+    /// - newest persistent data record
+    /// - address of first empty slot
+    fn persistent_data_scan(&self) -> (Option<RawPersistentData>, Option<u32>) {
+        let mut best: Option<RawPersistentData> = None;
+        let mut empty_slot: Option<u32> = None;
+        for i in 0..SECTOR_SIZE_BYTES / HF_PERSISTENT_DATA_STRIDE {
+            let addr = (i * HF_PERSISTENT_DATA_STRIDE) as u32;
+            let mut data = RawPersistentData::new_zeroed();
+            self.qspi.read_memory(addr, data.as_bytes_mut());
+            if data.is_valid()
+                && best.map(|b| b.epoch < data.epoch).unwrap_or(true)
+            {
+                best = Some(data)
+            }
+            if empty_slot.is_none()
+                && data.as_bytes().iter().all(|b| *b == 0xFF)
+            {
+                empty_slot = Some(addr);
+            }
+        }
+        (best, empty_slot)
+    }
+
+    fn raw_sector0_erase(&mut self) -> Result<(), HfError> {
+        set_and_check_write_enable(&self.qspi)?;
+        self.qspi.sector_erase(0);
+        poll_for_write_complete(&self.qspi, Some(1));
+        Ok(())
+    }
+
+    fn write_raw_persistent_data_to_addr(
+        &mut self,
+        addr: Option<u32>,
+        raw_data: &RawPersistentData,
+    ) -> Result<(), HfError> {
+        let addr = match addr {
+            Some(a) => a,
+            None => {
+                self.raw_sector0_erase()?;
+                0
+            }
+        };
+        Self::page_program_raw(&self.qspi, addr, raw_data.as_bytes())
+    }
+
+    /// Checks that the persistent data is consistent between the two flash ICs.
+    fn ensure_persistent_data_is_redundant(&mut self) -> Result<(), HfError> {
+        // This should only be called on startup, at which point we're always
+        // muxed to the SP.
```

Comment:
Is this comment true even if this task panics and restarts while the host is in A0?

Reply:
Yes – in `Config::init` in this same file, we ensure that the flash is muxed to the SP.

(Whether this is a good idea is questionable!  If the task restarts while we're in A0, that would be bad!)

Reply:
There is implicit coordination with the sequencer – we'd have a bad time if the `hf` task restarts while the host is reading its phase1 image from QSPI – but this PR doesn't make things any worse than before.  I added some comments to that effect in the relevant ticket (#994)